### ISSUE TITLE
NetManager topic de-registration

### ISSUE THREAD

**Issue by giobart** [oakestra/oakestra-net#74]
## Short
In some cases, the net manager de-registers immediately from the service updates. 

## Proposal
**Example**:
```
2022/11/14 11:31:05 Received HTTP request - /container/deploy
2022/11/14 11:31:05 ReqBody received : [123 34 112 105 100 34 58 49 50 54 56 56 44 34 115 101 114 118 105 99 101 78 97 109 101 34 58 34 99 108
105 101 110 116 46 112 101 114 99 111 109 109 46 99 108 105 101 110 116 46 100 101 112 108 111 121 34 44 34 105 110 115 116 97 110 99 101 78 1$
7 109 98 101 114 34 58 50 44 34 112 111 114 116 77 97 112 112 105 110 103 115 34 58 34 34 125]
2022/11/14 11:31:05 {12688 client.percomm.client.deploy 2  131.159.24.80 50103 0xc00025e300 0xc0003fe050 0xc00010e180}
2022/11/14 11:31:05 MQTT - publish to - service/deployed - the payload - {"appname":"client.percomm.client.deploy","status":"DEPLOYED","instan$
e_number":2,"nsip":"10.18.0.130","host_port":"50103","host_ip":"131.159.24.80"}
2022/11/14 11:31:05 [MQTT TABLE QUERY] sname: client.percomm.client.deploy
2022/11/14 11:31:05 MQTT - publish to - tablequery/request - the payload - {"sname":"client.percomm.client.deploy","sip":""}
2022/11/14 11:31:05 waiting for table query client.percomm.client.deploy
2022/11/14 11:31:05 MQTT - Received mqtt table query message: {"app_name": "client.percomm.client.deploy", "instance_list": [{"cluster_id": "6$
720dcb7da606aeff413bd5", "instance_ip": "10.30.0.13", "instance_number": 2, "worker_id": "6372110f665f0c3b79ca0234", "namespace_ip": "10.18.0.$
30", "host_ip": "131.159.24.80", "host_port": 50103, "service_ip": [{"Address": "10.30.0.0", "IpType": "RR"}, {"IpType": "instance_ip", "Addre$
s": "10.30.0.13"}]}]}
2022/11/14 11:31:05 TableQuery response - notifying a channel regarding client.percomm.client.deploy
2022/11/14 11:31:05 MQTT - Subscribed to jobs/client.percomm.client.deploy/updates_available
2022/11/14 11:31:05 Response to /container/deploy:  {client.percomm.client.deploy 10.18.0.130}
2022/11/14 11:31:05 self destruction timeout started for job client.percomm.client.deploy
2022/11/14 11:31:05 [MQTT TABLE QUERY] sip: 10.30.100.1
2022/11/14 11:31:05 MQTT - publish to - tablequery/request - the payload - {"sname":"","sip":"10.30.100.1"}
2022/11/14 11:31:05 waiting for table query 10.30.100.1
2022/11/14 11:31:05 MQTT - Received mqtt table query message: {"app_name": "pipeline.percomm.primary.deploy", "instance_list": [{"cluster_id":
"63720dcb7da606aeff413bd5", "instance_ip": "10.30.0.6", "instance_number": 0, "worker_id": "63720e27665f0c3b79c9fc9d", "namespace_ip": "10.18.$
.2", "host_ip": "131.159.25.19", "host_port": 50103, "service_ip": [{"Address": "10.30.100.1", "IpType": "RR"}, {"IpType": "instance_ip", "Add$
ess": "10.30.0.6"}]}]}
2022/11/14 11:31:05 TableQuery response - notifying a channel regarding 10.30.100.1
2022/11/14 11:31:05 MQTT - Subscribed to jobs/pipeline.percomm.primary.deploy/updates_available
2022/11/14 11:31:05 self destruction timeout started for job pipeline.percomm.primary.deploy
2022/11/14 11:31:05 MQTT - Subscribed to jobs/10.30.100.1/updates_available
2022/11/14 11:31:05 self destruction timeout started for job 10.30.100.1
2022/11/14 11:31:15 De-registering from 10.30.100.1
2022/11/14 11:31:15 MQTT - publish to - interest/remove - the payload - {"appname":"10.30.100.1"}
2022/11/14 11:31:15 De-registering from pipeline.percomm.primary.deploy
2022/11/14 11:31:15 MQTT - publish to - interest/remove - the payload - {"appname":"pipeline.percomm.primary.deploy"}
ERROR-2022/11/14 11:31:15 ProxyTunnel.go:553: Unable to resolve remote addr: address -1: invalid port
ERROR-2022/11/14 11:31:15 ProxyTunnel.go:553: Unable to resolve remote addr: address -1: invalid port
```

**Steps**
1: We deploy the service:  client.percomm.client.deploy with serviceip 10.30.0.0 and instaceip 10.30.0.13
2: Then the service requires a table query for service pipeline.percomm.primary.deploy with RRIP 10.30.100.1 
3: The self-destruction timeout towards 10.30.100.1 starts
4: The traffic flows and the application works as desired
5: `De-registering from 10.30.100.1` even if the application is still sending traffic to 10.30.100.1
6: As a result ERROR-2022/11/14 11:31:15 ProxyTunnel.go:553: Unable to resolve remote addr: address -1: invalid port
7: No further table queries!

So we have two different problems:
- Undesired de-registration
- No table query towards that route. Something is preventing the net manager to handle the table query

**Setup**:
client.percomm.client.deploy -> Intel NUC VM hosting only NodeEngine and NetManager
pipeline.percomm.primary.deploy  -> AMD64 VM hosting Root Orch + Cluster Orch + Node Engine 

the machines are mutually reachable (as proof, the application initially works)

**Possible causes**:
-Something is preventing the self-destruction timer from properly "ticking"

## Solution 
--

## Status
searching for a solution

## Checklist
- [x] Discussed
- [x] Solved
- [x] Tested

**Comment by nitindermohan** [comment]
@giobart So the issue was solved by adding a mutex lock?

**Comment by giobart** [comment]
@nitindermohan I solved the issue by

Hitting the table cache and triggering the table entry usage event for every incoming packet: node-net-manager/proxy/ProxyTunnel.go l241:248
```
		tableEntryList := proxy.environment.GetTableEntryByServiceIP(ipv4.DstIP)
 		if len(tableEntryList) < 1 {
 			logger.DebugLogger().Printf("No entries found for this service IP: %s", ipv4.DstIP.String())
 			return nil
 		}
```

With this solution we had a race condition if incoming packets were coming during an undeployment due to the concurrent event map RW operation. Therefore I added a RWLock in node-net-manager/events/EventManager.go
```
var rwlock sync.RWMutex
...
func (e *Events) Emit(event Event) {
 	rwlock.RLock()
 	defer rwlock.RUnlock()
...
```

Then I came to understand that using an hash map for the proxy cache is not very efficient. Therefore for the proxy cache I updated the implementation with an array large as the number of ports (node-net-manager/proxy/proxyCache.go line 33). This uses some extra memory but saves CPU cycles which is preferable during proxy conversions. 
```
+		cache:                 make(map[int]ConversionList),
- 		cache:                 make([]ConversionList, 65535),
```
